src/main/java/bio/terra/service/job/JobService.java

```diff
@@ -323,43 +330,49 @@ public List<JobModel> enumerateJobs(
   private List<FlightState> getAccessibleFlights(
       int offset, int limit, FlightFilter filter, AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    int start = 0;
-    int end = offset + limit;
+
+    String nextPageToken = null;
     HashMap<String, List<String>> roleMap = new HashMap<>();
     ArrayList<FlightState> flightStateList = new ArrayList<>();
-    while (flightStateList.size() < offset + limit) {
-      List<FlightState> filterResults = stairway.getFlights(start, end, filter);
-      if (filterResults.size() == 0) {
+    int numTimesQueried = 0;
+    while (flightStateList.size() < offset + limit && numTimesQueried < MAX_FLIGHT_SEARCH_QUERIES) {
+      logger.info("Getting flight batch {} with nextPageToken {}", numTimesQueried, nextPageToken);
+      FlightEnumeration filterResults =
+          stairway.getFlights(nextPageToken, FLIGHT_SEARCH_BATCH_SIZE, filter);
+      if (filterResults.getFlightStateList().size() == 0) {
         break;
       }
-      filterResults.forEach(
-          flightState -> {
-            if (userLaunchedFlight(flightState, userReq)) {
+
+      for (var flightState : filterResults.getFlightStateList()) {
+        if (userLaunchedFlight(flightState, userReq)) {
+          flightStateList.add(flightState);
+        } else {
+          FlightMap inputParameters = flightState.getInputParameters();
+          IamResourceType resourceType =
+              inputParameters.get(JobMapKeys.IAM_RESOURCE_TYPE.getKeyName(), IamResourceType.class);
+          String resourceId =
+              inputParameters.get(JobMapKeys.IAM_RESOURCE_ID.getKeyName(), String.class);
+          IamAction action =
+              inputParameters.get(JobMapKeys.IAM_ACTION.getKeyName(), IamAction.class);
+          if (resourceType != null & resourceId != null && action != null) {
+            String key = String.format("%s-%s", resourceType, resourceId);
+            List<String> userRoles = roleMap.get(key);
+            if (userRoles == null) {
+              userRoles = samService.listActions(userReq, resourceType, resourceId);
```

Comment:
Not a question to block the PR.
Does SAM have a way of providing _all_ of the listActions for a given user?  Calling this iteratively seems like it could really slow things down.  If we had a way to ask SAM, for the given user, give me all of the resourceTypes and resource IDs the user has access to, we might be able to invert this query and loop.

Reply:
> Does SAM have a way of providing all of the listActions for a given user?

Saman raised this question to Identiteam, because we thought there was a SAM endpoint that provided this.  Unfortunately, it does not look to be the case, but Identiteam might be open to making the change for us if we requested.  More info in thread:

https://broadinstitute.slack.com/archives/C03GMG4DUSE/p1663099459055089